nifi-nar-bundles/nifi-social-media-bundle/nifi-twitter-processors/src/main/java/org/apache/nifi/processors/twitter/TweetStreamService.java

```diff
@@ -0,0 +1,222 @@
+    }
+
+    private Long calculateBackoffDelay() {
+        long backoff = backoffMultiplier * backoffTime;
+        long delay = (backoff > maximumBackoff) ? maximumBackoff : backoff;
+        return delay;
+    }
+
+    private void scheduleStartStreamWithBackoff() {
+        // use exponential(by factor of 2) backoff in scheduling the next TweetStreamStarter
+        if (attemptCounter >= backoffAttempts) {
+            throw new ProcessException(String.format("Connection failed after maximum attempts [%d]", attemptCounter));
+        }
+        attemptCounter += 1;
+        long delay = calculateBackoffDelay();
+        backoffMultiplier *= 2;
+        logger.info("Scheduling new stream connection after delay [{} s]", delay);
+        executorService.schedule(new TweetStreamStarter(), delay, TimeUnit.SECONDS);
+    }
+
+    private void resetBackoff() {
+        attemptCounter = 0;
+        backoffMultiplier = 1L;
+    }
+
+    private class TweetStreamStarter implements Runnable {
+        @Override
+        public void run() {
+            try {
+                if (endpoint.equals(StreamEndpoint.SAMPLE_ENDPOINT)) {
+                    stream = api.tweets().sampleStream(expansions, tweetFields, userFields, mediaFields, placeFields, pollFields, backfillMinutes);
+                } else {
+                    stream = api.tweets().searchStream(expansions, tweetFields, userFields, mediaFields, placeFields, pollFields, backfillMinutes);
+                }
+                executorService.execute(new TweetStreamHandler());
+            } catch (ApiException e) {
+                stream = null;
+                logger.warn("Twitter Stream [{}] API connection failed: HTTP {}", endpoint.getEndpointName(), e.getCode(), e);
+                scheduleStartStreamWithBackoff();
+            } catch (Exception e) {
+                stream = null;
+                logger.warn("Twitter Stream [{}] connection failed", endpoint.getEndpointName(), e);
+                scheduleStartStreamWithBackoff();
+            }
+        }
+    }
+
+    private class TweetStreamHandler implements Runnable {
+        @Override
+        public void run() {
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(stream))) {
+                String tweetRecord = reader.readLine();
+                while (tweetRecord != null) {
+                    queue.offer(tweetRecord);
+
+                    // reset backoff multiplier upon successful receipt of a tweet
+                    resetBackoff();
+
+                    tweetRecord = reader.readLine();
+                }
+            } catch (IOException e) {
+                logger.info("Stream is closed or has stopped", e);
+            }
+            logger.info("Tweet stream processing completed");
+            scheduleStartStreamWithBackoff();
+        }
+    }
+
+    private Set<String> parseCommaSeparatedPropreties(final ProcessContext context, final PropertyDescriptor property) {
```

Comment:
Minor wording correction:
```suggestion
    private Set<String> parseCommaSeparatedProperties(final ProcessContext context, final PropertyDescriptor property) {
```

Reply:
Thanks for the catch, making the change